File: demo-task-executor-java/src/main/java/demo/executor/Executor.java
package demo.executor;

import java.time.Duration;

/**
 * Executor to run multiple {@link demo.executor.Task}.
 */
public interface Executor {

    /**
     * Submits a task to the executor.
     *
     * @param task the task to run.
     */
    void submit(Task<?> task);

    /**
     * Waits until all the submitted  tasks are finished.
     */
    void waitUntilFinished();

    /**
     * Stops the executor after finishing all previously submitted tasks.
     *
     * @return the executor work duration.
     */
    Duration stop();
}

File: demo-task-executor-kotlin/src/main/kotlin/demo/executor/Executor.kt
package demo.executor

import java.time.Duration

/**
 * Executor to run multiple [demo.executor.Task].
 */
interface Executor {

    /**
     * Submits a task to the executor.

     * @param task the task to run.
     */
    fun submit(task: Task<*>)

    /**
     * Waits until all the submitted  tasks are finished.
     */
    fun waitUntilFinished()

    /**
     * Stops the executor after finishing all previously submitted tasks.

     * @return the executor work duration.
     */
    fun stop(): Duration
}
